crates/jstz_engine/src/script.rs

```diff
@@ -0,0 +1,137 @@
+    where
+        S: InCompartment<C> + CanAlloc,
+    {
+        let mut source = mozjs::rust::transform_str_to_source_text(script);
+        let options = unsafe {
+            CompileOptionsWrapper::new(cx.as_raw_ptr(), path.to_str().unwrap(), 1)
+        };
+
+        let script = unsafe { Compile1(cx.as_raw_ptr(), options.ptr, &mut source) };
+
+        Some(Self {
+            script: NonNull::new(script)?,
+            marker: PhantomData,
+        })
+    }
+
+    /// Evaluates a script and returns its return value
+
+    // TODO(https://linear.app/tezos/issue/JSTZ-210):
+    // Add support for error handling / exceptions instead of using `Option`
+    // TODO(https://linear.app/tezos/issue/JSTZ-211):
+    // TODO: `JSVal` is not safe, we should return a safe wrapper instead
+    pub fn evaluate<'b, S>(&self, cx: &'b mut Context<S>) -> Option<JSVal>
+    where
+        S: InCompartment<C> + CanAlloc,
+        'a: 'b,
+    {
+        // TODO(https://linear.app/tezos/issue/JSTZ-196):
+        // Remove this once we have a proper way to root values
+        rooted!(in(unsafe { cx.as_raw_ptr() }) let mut rval = UndefinedValue());
+        rooted!(in(unsafe { cx.as_raw_ptr() }) let mut rooted_script = unsafe { self.as_raw_ptr() });
+
+        if unsafe {
+            JS_ExecuteScript(
+                cx.as_raw_ptr(),
+                rooted_script.handle_mut().into(),
+                rval.handle_mut().into(),
+            )
+        } {
+            Some(rval.get())
+        } else {
+            None
+        }
+    }
+}
+
+impl<'a, C: Compartment> AsRawPtr for Script<'a, C> {
+    type Ptr = *mut JSScript;
+
+    unsafe fn as_raw_ptr(&self) -> Self::Ptr {
+        self.script.as_ptr()
+    }
+}
+
+#[cfg(test)]
+mod test {
+
+    use std::path::PathBuf;
+
+    use mozjs::rust::{JSEngine, Runtime};
+
+    use crate::{compartment, context::Context, script::Script};
+
+    #[test]
+    fn test_compile_and_evaluate() {
+        // Initialize the JS engine.
+        let engine = JSEngine::init().unwrap();
+        let rt = Runtime::new(engine.handle());
+        let rt_cx = &mut Context::from_runtime(&rt);
+
+        // Enter a new realm to evaluate the script in.
+        let mut cx = rt_cx.new_realm().unwrap();
+
+        // Some example source in a string.
+        let filename = PathBuf::from("inline.js");
+        let source: &'static str = "40 + 2";
+
+        // Compile the script
+        let script = Script::compile(&filename, source, &mut cx).unwrap();
+
+        // TODO(https://linear.app/tezos/issue/JSTZ-196):
+        // Remove once we have a proper way of rooting things.
+        // The script is rooted in the context in `eval`, but this doesn't work due to lifetimes.
+        // So we need to transmute it here.
+        let rooted_script: Script<'_, compartment::Ref<'_>> =
+            unsafe { std::mem::transmute(script) };
+
+        // Evaluate the script
+        let res = rooted_script.evaluate(&mut cx);
```

Comment:
I wonder if you thought about Script holding a `*mut JSContext` since, in most cases,

Reply:
Not always, scripts are nothing more than compiled bits of bytecode. They require a *context* in order to execute. We should also be weary of implicitly storing contexts since they are the capability for doing anything in SpiderMonkey (and we almost should never implicitly copy / clone a context)